category/vm/compiler/ir/x86/emitter.cpp

```diff
@@ -7928,8 +7969,36 @@ namespace monad::vm::compiler::native
         }
     }
 
+    // Performs leading zeros operation on array of operands. Assumes that the
+    // operands are ordered from least significant to most significant.
+    // The current implementation optimizes for the case where the input value
+    // is uniformly distributed. When it's uniformly distributed, the chance
+    // that the most significant 64-bit word == 0 is 1 / 2^64. We take advantage
+    // of this with a branched approach that will only compute the bit width of
+    // the most significant word when it's non-zero (which the branch predictor
+    // will predict correctly most of the time).
+    template <typename T, size_t N>
+    void Emitter::array_leading_zeros(std::array<T, N> const &arr)
+    {
+        auto const end_lbl = as_.newLabel();
+        for (size_t i = N; i >= 1; --i) {
+            auto const word_offset = static_cast<int32_t>(64 * (N - i));
+            // Compute number of leading zeros. CF == 1 iff arr[i] == 0
+            as_.lzcnt(x86::rax, arr[i - 1]);
+            if (word_offset != 0) {
+                // Leave flags unchanged
+                as_.lea(x86::eax, x86::ptr(x86::eax, word_offset));
+            }
+            // If arr[i - 1] != 0, jump to end_lbl
+            as_.jnc(end_lbl);
+        }
+        as_.bind(end_lbl);
+    }
+
     // Performs byte_width operation on array of operands. Assumes that the
     // operands are ordered from least significant to most significant.
+    // This implementation assumes that the input value is biased towards
+    // smaller values.
```

Comment:
I think this comment was supposed to belong to `array_leading_zeros`

Reply:
You're right that this comment is a bit unclear. What it's trying to say is that `array_leading_zeros` and `array_byte_width` use 2 different approaches (one branchless, one branchful) for calculating something that are more-or-less equivalent, with each approach optimizing for different inputs:
- branchful: variable latency, returning faster for wider inputs (executing only `lzcnt, jnc` if 256-bit wide), such as is the case when they are uniformly distributed.
- branchless: always the same latency, no early exit for large operands so better in the case where the input won't occupy the full 256 bits such as the exponent of the `EXP` operation.

I'll rewrite the comment to make that clearer.